From patchwork Fri Dec  7 16:23:15 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Ahmed Abd El Mawgood <ahmedsoliman@mena.vt.edu>
X-Patchwork-Id: 10718509
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id AE96815A6
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri,  7 Dec 2018 16:24:06 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 97A6E28C2F
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri,  7 Dec 2018 16:24:06 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 86A3F2DDAB; Fri,  7 Dec 2018 16:24:06 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 0559C28C2F
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri,  7 Dec 2018 16:24:05 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1726132AbeLGQX7 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 7 Dec 2018 11:23:59 -0500
Received: from mail-wr1-f66.google.com ([209.85.221.66]:35936 "EHLO
        mail-wr1-f66.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726121AbeLGQX7 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 7 Dec 2018 11:23:59 -0500
Received: by mail-wr1-f66.google.com with SMTP id u3so4370826wrs.3
        for <kvm@vger.kernel.org>; Fri, 07 Dec 2018 08:23:58 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=mena-vt-edu.20150623.gappssmtp.com; s=20150623;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=0ZLMRK1sa14ZeWEcMpGckEyj8BI4tsg92VSqXi44f2k=;
        b=z1TUJhC/ICXd7VUEA7Zc9hLkfQ9RfS+KHHrmxIjly7beF6KVxTS+6rzLEs2J2ur9MT
         coAAwfSGk1MfNZxQItV0Rs4baXK/aFwLaaSEyCCF6J7S15hMOLm+GOhMp+7apzqVxgwM
         bnJxdcRXzUNcbo9uCcR2I9VowdftyuzVRjmpU5hQCltb5ofH48xrbHUVDK4TDYl9Y0Cy
         uO3pQ7tR4tqtSSw3Sa+zF+Fg3YrvFFuTOlwhKh3xxlXBvBpF+PJ82xi5rjuKCf5eRrB9
         N8aWEjzoqw7tLtcNW0f1yqj2FEHkAYuC7zvnjjvGsnxvpQYEUwaW5ULomgRSbLR8ZoMH
         /JoQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=0ZLMRK1sa14ZeWEcMpGckEyj8BI4tsg92VSqXi44f2k=;
        b=XJtyp8Wj2fD5j1FyAXkQadJUt5LjYs/1OU4WW7k1QexXKWCrxUCkMXMNdTRoEvQoaX
         Szi1AHyZyIB/aXlROtBjZx3P3vznvoZ9+bIbhZ7X8eWd4bsjJy69pPDSkSYZ3giCTjb2
         ObNbNrYVlWE1mbRWd0+ihxS+YnAeJP9GABn5XBEAdvl6rsz+7n4S7fp0Gargmyu38XCf
         gejm1JwC/JcGbmZNZrYPRNiZOZ5l8a7TQxx24mOmevb8hYP5/4tlL22qUjIYb5yq6TFP
         oPKEz6jBfdYFDyH2TcUEip3gXK3dvdMFBTWOWA8MaPAklPMqkccUG6+vefCk9hiQjbEh
         QKhw==
X-Gm-Message-State: AA+aEWYBIhFFCYvZxrbPvInCUNMS57/KFN706CI86/p5aIasAAPcRTrp
        CPD/tq+1nAEZ/7VwVaMe5ni7tDqPWDU=
X-Google-Smtp-Source: 
 AFSGD/UkT9Ta/PzidJmaQ+Am56ke5yurCQxQ2pToh570695xGXOPEh8uXHqsdnU0qD0vzyXjJMdsrA==
X-Received: by 2002:adf:8228:: with SMTP id 37mr2289815wrb.160.1544199837389;
        Fri, 07 Dec 2018 08:23:57 -0800 (PST)
Received: from localhost.localdomain ([156.213.98.90])
        by smtp.gmail.com with ESMTPSA id
 a1sm2378073wrw.76.2018.12.07.08.23.54
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Fri, 07 Dec 2018 08:23:56 -0800 (PST)
From: Ahmed Abd El Mawgood <ahmedsoliman@mena.vt.edu>
To: Paolo Bonzini <pbonzini@redhat.com>, rkrcmar@redhat.com,
        Jonathan Corbet <corbet@lwn.net>,
        Thomas Gleixner <tglx@linutronix.de>,
        Ingo Molnar <mingo@redhat.com>, Borislav Petkov <bp@alien8.de>,
        hpa@zytor.com, x86@kernel.org, kvm@vger.kernel.org,
        linux-doc@vger.kernel.org, linux-kernel@vger.kernel.org,
        ahmedsoliman0x666@gmail.com, ovich00@gmail.com,
        kernel-hardening@lists.openwall.com, nigel.edwards@hpe.com,
        Boris Lukashev <blukashev@sempervictus.com>,
        Igor Stoppa <igor.stoppa@gmail.com>
Cc: Ahmed Abd El Mawgood <ahmedsoliman@mena.vt.edu>
Subject: RESEND [PATCH 10/10] KVM: Log ROE violations in system log
Date: Fri,  7 Dec 2018 18:23:15 +0200
Message-Id: <20181207162315.6004-1-ahmedsoliman@mena.vt.edu>
X-Mailer: git-send-email 2.19.2
In-Reply-To: <20181207124803.10828-1-ahmedsoliman@mena.vt.edu>
References: 
MIME-Version: 1.0
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

I am absolutely sorry, I had some modifications that I forgot to commit
before I send. so please use this one patch 10/10 instead of the last
one.

Signed-off-by: Ahmed Abd El Mawgood <ahmedsoliman@mena.vt.edu>
---
 virt/kvm/kvm_main.c    |  3 ++-
 virt/kvm/roe.c         | 26 ++++++++++++++++++++++++++
 virt/kvm/roe_generic.h |  3 ++-
 3 files changed, 30 insertions(+), 2 deletions(-)

diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index c3a21d3bc8..761cb7561a 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -1870,13 +1870,14 @@ static u64 roe_gfn_to_hva(struct kvm_memory_slot *slot, gfn_t gfn, int offset,
 	addr = __gfn_to_hva_many(slot, gfn, NULL, false);
 	return addr;
 }
+
 static int __kvm_write_guest_page(struct kvm_memory_slot *memslot, gfn_t gfn,
 			          const void *data, int offset, int len)
 {
 	int r;
 	unsigned long addr;
-
 	addr = roe_gfn_to_hva(memslot, gfn, offset, len);
+	kvm_roe_check_and_log(memslot, gfn, data, offset, len);
 	if (kvm_is_error_hva(addr))
 		return -EFAULT;
 	r = __copy_to_user((void __user *)addr + offset, data, len);
diff --git a/virt/kvm/roe.c b/virt/kvm/roe.c
index 6555838f0c..01362f0fca 100644
--- a/virt/kvm/roe.c
+++ b/virt/kvm/roe.c
@@ -76,6 +76,32 @@ void kvm_roe_free(struct kvm_memory_slot *slot)
 	kvfree(slot->prot_list);
 }
 
+static void kvm_warning_roe_violation(u64 addr, const void *data, int len)
+{
+	int i;
+	const char *d = data;
+	char *buf = kvmalloc(len * 3 + 1, GFP_KERNEL);
+
+	for (i = 0; i < len; i++)
+		sprintf(buf+3*i, " %02x", d[i]);
+	pr_warn("ROE violation:\n");
+	pr_warn("\tAttempt to write %d bytes at address 0x%08llx\n", len, addr);
+	pr_warn("\tData: %s\n", buf);
+	kvfree(buf);
+}
+
+void kvm_roe_check_and_log(struct kvm_memory_slot *memslot, gfn_t gfn,
+		const void *data, int offset, int len)
+{
+	if (!memslot)
+		return;
+	if (!gfn_is_full_roe(memslot, gfn))
+		return;
+	if (!kvm_roe_check_range(memslot, gfn, offset, len))
+		return;
+	kvm_warning_roe_violation((gfn << PAGE_SHIFT) + offset, data, len);
+}
+
 static void kvm_roe_protect_slot(struct kvm *kvm, struct kvm_memory_slot *slot,
 		gfn_t gfn, u64 npages, bool partial)
 {
diff --git a/virt/kvm/roe_generic.h b/virt/kvm/roe_generic.h
index f1ce4a8aec..6c5f0cf381 100644
--- a/virt/kvm/roe_generic.h
+++ b/virt/kvm/roe_generic.h
@@ -14,5 +14,6 @@ void kvm_roe_free(struct kvm_memory_slot *slot);
 int kvm_roe_init(struct kvm_memory_slot *slot);
 bool kvm_roe_check_range(struct kvm_memory_slot *slot, gfn_t gfn, int offset,
 		int len);
-
+void kvm_roe_check_and_log(struct kvm_memory_slot *memslot, gfn_t gfn,
+		const void *data, int offset, int len);
 #endif
